File: hashset/cxx/HashSet.cxx
#include "HashSet.h"

HashSet::HashSet(): primeIdx(0), _size(0)
{
	table.resize(primes[primeIdx]);
}

int HashSet::primes[32] = {
	53, 97, 193, 389, 769, 1543, 3079, 6151, 12289, 24593, 49157, 98317,
	196613, 393241, 786433, 1572869, 3145739, 6291469, 12582917, 25165843,
	50331653, 100663319, 201326611, 402653189, 805306457, 1610612741
};

int HashSet::myhash(const int& x) const
{
	int hashVal = x;
	hashVal %= table.size();
	if(hashVal < 0)
		hashVal += table.size();
	return hashVal;
}

void HashSet::clear()
{
	_size = 0;
	for(int i = 0; i < table.size(); i++)
		table[i].clear();
}

bool HashSet::contains(const int& x) const
{
	const std::list<int>& theList = table[myhash(x)];
	return std::find(theList.begin(), theList.end(), x) != theList.end();
}

bool HashSet::erase(const int& x)
{
	std::list<int>& theList = table[myhash(x)];
	std::list<int>::iterator it = std::find(theList.begin(), theList.end(), x);

	if(it == theList.end())
		return false;
	theList.erase(it);
	_size--;
	return true;
}

bool HashSet::insert(const int& x)
{
	std::list<int>& theList = table[myhash(x)];
	if(std::find(theList.begin(), theList.end(), x) != theList.end())
		return false;
	theList.push_back(x);
	if(++_size > table.size())
		this->rehash();
	return true;
}

void HashSet::rehash()
{
	std::vector<std::list<int> > oldTable( primes[++primeIdx] );
	table.swap(oldTable);
	_size = 0;
	for(int i = 0; i < oldTable.size(); i++)
	{
		std::list<int>::iterator it = oldTable[i].begin();
		while(it != oldTable[i].end())
			this->insert(*it++);
	}
}

File: hashset/py3/HashSet.py

class HashSet(object):
    def __init__(self):
        self.primes = [53, 97, 193, 389, 769, 1543, 3079, 6151, 12289,
                24593, 49157, 98317, 196613, 393241, 786433, 1572869,
                3145739, 6291469, 12582917, 25165843, 50331653,
                100663319, 201326611, 402653189, 805306457, 1610612741]
        self.primeIdx = 0
        self.table = [[] for _ in range(self.primes[self.primeIdx])]
        self.size = 0
    
    def __len__(self):
        return self.size

    def add(self, value):
        cell = self.table[hash(value) % len(self.table)]
        try:
            cell.index(value)
        except ValueError:
            self.size += 1
            cell.append(value)
            if self.size > self.primes[self.primeIdx]:
                self.rehash()

    def remove(self, value):
        cell = self.table[hash(value) % len(self.table)]
        try:
            cell.remove(value)
            self.size -= 1
        except ValueError:
            raise KeyError(value)

    def __contains__(self, value):
        cell = self.table[hash(value) % len(self.table)]
        try:
            cell.index(value)
            return True
        except ValueError:
            return False

    def discard(self, value):
        cell = self.table[hash(value) % len(self.table)]
        try:
            cell.remove(value)
            self.size -= 1
        except ValueError:
            pass

    def clear(self):
        self.table = [[]] * self.primes[self.primeIdx]
        self.size = 0

    def rehash(self):
        oldTable = self.table
        self.primeIdx += 1
        self.table = [[] for _ in range(self.primes[self.primeIdx])]
        self.size = 0
        for cell in oldTable:
            for value in cell:
                self.add(value)
